### ISSUE TITLE
FXCop Analyzer CA1825 raises a warning eventhough it is suppressed in GlobalSupression.cs

### ISSUE THREAD

**Issue by vsfeedback** [dotnet/roslyn-analyzers#3587]
_This issue has been moved from [a ticket on Developer Community](https://developercommunity.visualstudio.com/content/problem/1012376/fxcop-analyzer-ca1825-raises-a-warning-eventhough-2.html)._

---
1. Added `Microsoft.CodeAnalysis.FxCop Analyzers` to `NuGet.Client` repo.
1. Supressed all the existing violations using `Analyze -> BuildAndSupressActiveIssues -> For Solution` option in Visual Studio
1. Observed that FXCop Analyzer `CA1825` raises a warning eventhough it is suppressed in `GlobalSupression.cs`

Observed the same behavior for following analyzers
1. CA1062: Validate arguments of public methods
2. CA1303: Do not pass literals as localized parameters
3. CA1801: Review unused parameters
4. CA1822: Mark members as static
5. CA1305: Specify IFormatProvider
6. CA1031: Do not catch general exception types

Our repo has `TreatWarningsAsErrors` to `true`. These warnigs are causing build failures locally and in CI/CD. We have changed the severity of these 7 rules to `Info` to unblock our builds.

---
#### Visual Studio Feedback System on 4/29/2020, 08:38 PM: 

<p>We have directed your feedback to the appropriate engineering team for further evaluation. The team will review the feedback and notify you about the next steps.</p>

#### Visual Studio Feedback System on 4/30/2020, 09:53 AM: 

<p>Thank you for sharing your feedback! Our teams prioritize action on product issues with broad customer impact. See details at: <a target='_blank' href="https://docs.microsoft.com/en-us/visualstudio/ide/report-a-problem?view=vs-2019#faq">https://docs.microsoft.com/en-us/visualstudio/ide/report-a-problem?view=vs-2019#faq</a>. In case you need answers to common questions or need assisted support, be sure to use <a target='_blank' href="https://visualstudio.microsoft.com/vs/support/">https://visualstudio.microsoft.com/vs/support/</a>. We’ll keep you posted on any updates to this feedback.</p>

---
### Original Solutions
(no solutions)

**Comment by mavasani** [comment]
@kartheekp-ms Is this the same issue as https://github.com/dotnet/roslyn-analyzers/issues/3629? Are you still affected by this?

**Comment by kartheekp-ms** [comment]
@mavasani - No, this is a different issue. We are blocked by this issue and hence marked the severity of above said analyzers to `Info` to avoid build failures as we treat warnings as errors.

I have changed the severity of `CA1825` to `Warning` in my local repo and analyzer reported error as shown below even though we have an entry in the [`GlobalSuppressions.cs`](https://github.com/NuGet/NuGet.Client/blob/dev/src/NuGet.Clients/NuGet.CommandLine/GlobalSuppressions.cs#L77) to suppress warning reported [here](https://github.com/NuGet/NuGet.Client/blob/dev/src/NuGet.Clients/NuGet.CommandLine/Common/MSBuildProjectSystem.cs#L434).

![image](https://user-images.githubusercontent.com/52756182/85763728-ccddef80-b6c9-11ea-8eba-ada8dafb65ba.png)

**Comment by sharwell** [comment]
> ... even though we have an entry in the GlobalSuppressions.cs to suppress warning ... 

My first guess would be that the syntax used in the global suppression is not correct. @mavasani might have a link to the expected syntax.

**Comment by sharwell** [comment]
Starting with Visual Studio 2019 version 16.7 Preview 2, a new analyzer (https://github.com/dotnet/roslyn/pull/44287) will report cases where the target of a global suppression cannot be found. The first thing I would try is opening the solution in this version and looking for reports of IDE0076.

**Comment by kartheekp-ms** [comment]
> My first guess would be that the syntax used in the global suppression is not correct. @mavasani might have a link to the expected syntax. I would expect the target to just be this:
> 
> ```
> M:NuGet.Common.MSBuildProjectSystem.GetProject(System.String)
> ```

Thanks for the suggestion. That was interesting and good to know because removing `~` char as suggested above removed the error from error list window. We have suppressed all the current violations (baselining) by selecting **Analyze > Build and Suppress Active Issues** on the menu bar in VS IDE as per the [doc](https://docs.microsoft.com/en-us/visualstudio/code-quality/use-roslyn-analyzers?view=vs-2019#suppress-violations). Is there any remote possibility of a bug in that logic that appended extra `~` char to the `Target ` attribute ?

**Comment by kartheekp-ms** [comment]
> Starting with Visual Studio 2019 version 16.7 Preview 2, a new analyzer ([dotnet/roslyn#44287](https://github.com/dotnet/roslyn/pull/44287)) will report cases where the target of a global suppression cannot be found. The first thing I would try is opening the solution in this version and looking for reports of IDE0076 in GlobalSuppressions.cs.

That helps. I can see 13 warnings raised by `IDE0076` analyzer when I opened the solution in VS 2019 Int Preview (16.7 Preview 4). I found around 35 additional errors from `AD0001` analyzer. please see attachment for errors with stack trace. Looks like an issue in `IDE0076 and IDE0077` analyzers.
[AD0001 errors.txt](https://github.com/dotnet/roslyn-analyzers/files/4833849/AD0001.errors.txt)

**Comment by mavasani** [comment]
Thanks @kartheekp-ms. I filed https://github.com/dotnet/roslyn/issues/45465 for the AD0001 diagnostics.

**Comment by mavasani** [comment]
> Is there any remote possibility of a bug in that logic that appended extra ~ char to the Target attribute 

No, that is not a bug. Suppress message decoder in the compiler layer considers `~` as an optional starting character, so presence or absence of `~` has no effect on the suppression target.

**Comment by kartheekp-ms** [comment]
> > Is there any remote possibility of a bug in that logic that appended extra ~ char to the Target attribute
> 
> No, that is not a bug. Suppress message decoder in the compiler layer considers `~` as an optional starting character, so presence or absence of `~` has no effect on the suppression target.

I am curious to know why the error reported by `CA1825` analyzer has disappeared from the error list window once I removed `~` char at the end in the `Target` attribute as suggested by @sharwell. `GlobalSuppressions.cs` files are auto-generated when we did the baseline.

**Comment by sharwell** [comment]
@mavasani is talking about the `~` at the beginning. The one I suspected was problematic was the one at the _end_.

**Comment by mavasani** [comment]
@sharwell is correct, I was only mentioning the prefix `~`. The target string should never end with a `~`. If you have a repro case where a VS IDE code fix or command is generating such an invalid suppress message attribute, then please file an issue on Roslyn repo.

**Comment by kartheekp-ms** [comment]
I filed dotnet/roslyn#45466 for the extra `~` char noticed at the end of `Target` in `GlobalSuppressions.cs` file

**Comment by Evangelink** [comment]
If I got things right the issue is only on roslyn's side so this ticket could be closed, am I missing something?

**Comment by mavasani** [comment]
Yes, indeed. There is no bug in the analyzer(s). Closing out.
